### PR TITLE
Add manual content scanner

### PR DESCRIPTION
**Pull request by jdgleaver**
## Description

This PR adds a simple manual content scanner - i.e. a 'dumb' scanner that matches all files (based on user configuration) without checking the databases. The interface looks like this:

Screenshot_2019-11-29_17-17-34

Screenshot_2019-11-29_17-18-08

The scan options are as follows:

- `Content Directory`: Select directory to scan. The starting path for this is configured via the normal `File Browser` path in `Settings > Directory`, so it is highly recommended to set this first before beginning a scanning session.

- `System Name`: This corresponds to the conventional 'database' setting, used to name the resultant playlist and to identify thumbnails. This has the following options:

    - `<Content Directory>`: The name of the content directory is used

    - `<Custom>`: Allows entry of a custom name

    - The user can also select the name of any installed database

- `Custom System Name`: This is the name used when `System Name` is set to `<Custom>` (it is otherwise ignored)

- `Core`: Allows a default core to be specified

- `File Extensions`: Allows user to filter scanned files by specifying a space-delimited list of allowed file extensions. Some further notes are required here:

    - If `File Extensions` is left blank *and* no `Core` is specified, *all* files are scanned

    - If `File Extensions` is left blank *and* a `Core` is specified, all files supported by the core (+ all archive files) are scanned

    - If `File Extensions` is specified, *only* files with matching extensions are matched. An example use case: The PCSX ReARMed core supports both `bin` and `cue` files, which means you end up with a mess of duplicates in the playlist by default. Setting `File Extensions` to `cue` excludes the `bin`s and keeps everything clean.

- `Overwrite Existing Playlist`: When enabled, any existing playlist content will be erased (useful if the user previously made a mistake, and wants to start over). When disabled, only content not already present in the playlist will be added (useful as a second pass to a conventional database scan, to pick up any files that were missed)
 
NOTE: I realise that this scanner will produce poor results for Arcade (MAME, FBNeo) content, due to the fact that the ROM files are all zips with meaningless names. I have some ideas about this, and will try to address the issue in a future PR.

### CODE REVIEW

**Comment by Tatsuya79** [comment]
Awesome! :+1: 

Yes for arcade roms (also software lists for home machines) many of us are still using scripts with a database (usually from [here](http://www.progettosnaps.net/dats/MAME/)).

File: tasks/task_manual_content_scan.c [hunk 1]
```diff
@@ -0,0 +1,354 @@
+/*  RetroArch - A frontend for libretro.
+ *  Copyright (C) 2011-2017 - Daniel De Matteis
+ *  Copyright (C) 2014-2017 - Jean-André Santoni
+ *  Copyright (C) 2016-2019 - Brad Parker
+ *
```
Comment (askotx) [inline tasks/task_manual_content_scan.c]: Add your name here too, it's your addition :)

**Comment by Tatsuya79** [comment]
In the qt scanner (manual, right click>"add folder") there's an option to scan extensions inside zips which is pretty nice.

File: tasks/task_manual_content_scan.c (hunk 1 above)
Comment (jdgleaver) [reply tasks/task_manual_content_scan.c]: Ha! Yes, you're right :)

I've added quite a few source files, but I've always just copy/pasted the header comment without really looking at it. I need to go through all of them and update the copyright text!

**Comment by jdgleaver** [comment]
@Tatsuya79 

> Yes for arcade roms (also software lists for home machines) many of us are still using scripts with a database (usually from here).

Yep, that's how I currently make my arcade playlists.

I think allowing the user to manually select a DAT file via the scanner interface is the way to go here - it's flexible and future proof. I need to have a look at the new XML parser that was recently added to RetroArch. If it's okay, this should be fairly easy to implement.

> In the qt scanner (manual, right click>"add folder") there's an option to scan extensions inside zips which is pretty nice.

Okay, that makes sense. I'll have a go at adding this next week.

> Also I noticed the playlist isn't visible directly into qt when using that new manual scanner. But it's there after a reboot.

Yeah... AFAIK, the interface between Qt and the regular menus is only one-directional: Qt can trigger an update of RetroArch, but not the other way around.

I honestly have no interest in the Qt side of things, so probably won't look into this... But if someone can implement a hook for triggering a Qt update (or point me to something that already exists), i can wire it up to the manual scan task callback function.

**Comment by fr500** [comment]
This is pretty smart I hope the normal scanner gets updated to be as smart.
Say scan a folder against a PARTICULAR DB

That way you can avoid the mess that is created when scanning some arcade sets.
I also had this playlist related suggestion:

https://app.moqups.com/6iQYsTXUaj/view

**Comment by realnc** [comment]
Thanks for working on this! Just tried it on Linux. Works well.

**Comment by jdgleaver** [comment]
@fr500 Yes, I agree that the database-based scanner could use a similar overhaul, and your 'add to (any) playlist' idea could be useful.

I won't have a huge amount of time for RetroArch development in the run-up to Christmas, and what time I do have is already accounted for - but I'll try to have a look at these additional things in the new year!

**Comment by twinaphex** [comment]
@hizzlekizzle has tested this and given it a thumbs up. So I think we can merge this then.

**Comment by fr500** [comment]
@jdgleaver a suggestion/idea
What about an option to use parent folder name for entry name.

For people using Amiga, DOS, etc it would help a lot I think.
For examples I have my DOS games like this

```
Commander Keen 5 - The Armageddon Machine (1991)
|-------KEEN5.exe
Commander Keen 6 - Aliens Ate My Baby Sitter! (1991)
|-------KEEN6.exe
```

I figure most people using DOS would do something like that.
Otherwise you'll end up with KEEN5, KEEN6 entries instead of the game name.

**Comment by fr500** [comment]
A second idea

image

In this menu we could have an entry to select Thumbnails, at least for Title and Snaps, and <wishful thinking> a submit thumbnails option </wishful thinking>

**Comment by jdgleaver** [comment]
@fr500 Yeah, these are good ideas.

> What about an option to use parent folder name for entry name.

I did already think about this, and I need to think about it quite a bit more... It would be a very useful feature, but it is absolutely non-trivial - i.e. when a subdirectory contains multiple files, how do you handle them such that they don't all end up with the same label?

I actually think that these 'unusual' cores (DOSBox, UAE, PrBoom, TyrQuake) would really benefit from a ScummVM-style 'id' file (the manual scanner works great with ScummVM if you drop a `.scummvm` file in each content subdirectory).

So you'd just have a user-nameable file with a core-specific extension, with a line of text inside specifying the actual content file - in your DOS example, it would be like this:

```
Commander Keen 5 - The Armageddon Machine (1991)
|-------KEEN5.exe
|-------Commander Keen 5 - The Armageddon Machine (1991).dosbox
````

...where `Commander Keen 5 - The Armageddon Machine (1991).dosbox` contains:

```
KEEN5.exe
```
...and then someone could set up a repo like the one we have here: https://github.com/RobLoach/libretro-database-scummvm/tree/master/games

This requires some initial setup by the user, but some kind of standard like this across all cores that don't use single ROM files would make scanning far more consistent

**Pull request merged**